File: Source/AlphaTab/Rendering/Glyphs/BeatGlyphBase.cs
using System;
using AlphaTab.Model;

namespace AlphaTab.Rendering.Glyphs
{
    internal class BeatGlyphBase : GlyphGroup
    {
        public BeatContainerGlyph Container { get; set; }

        public BeatGlyphBase()
            : base(0, 0)
        {
        }

        public override void DoLayout()
        {
            // left to right layout
            var w = 0f;
            if (Glyphs != null)
            {
                for (int i = 0, j = Glyphs.Count; i < j; i++)
                {
                    var g = Glyphs[i];
                    g.X = w;
                    g.Renderer = Renderer;
                    g.DoLayout();
                    w += g.Width;
                }
            }

            Width = w;
        }

        protected void NoteLoop(Action<Note> action)
        {
            for (var i = Container.Beat.Notes.Count - 1; i >= 0; i--)
            {
                action(Container.Beat.Notes[i]);
            }
        }
    }
}

File: packages/alphatab/src/rendering/glyphs/BeatGlyphBase.ts
import type { BeatSubElement } from '@src/model/Beat';
import type { Note } from '@src/model/Note';
import type { BeatContainerGlyph } from '@src/rendering/glyphs/BeatContainerGlyph';
import type { Glyph } from '@src/rendering/glyphs/Glyph';
import { ElementStyleHelper } from '@src/rendering/utils/ElementStyleHelper';
import { GlyphGroup } from '@src/rendering/glyphs/GlyphGroup';
import type { ICanvas } from '@src/platform/ICanvas';

export class BeatGlyphBase extends GlyphGroup {
    private _effectGlyphs: Glyph[] = [];
    private _normalGlyphs: Glyph[] = [];

    public container!: BeatContainerGlyph;
    public computedWidth: number = 0;

    public constructor() {
        super(0, 0);
    }

    public override doLayout(): void {
        // left to right layout
        let w: number = 0;
        if (this.glyphs) {
            for (let i: number = 0, j: number = this.glyphs.length; i < j; i++) {
                const g: Glyph = this.glyphs[i];
                g.x = w;
                g.renderer = this.renderer;
                g.doLayout();
                w += g.width;
            }
        }
        this.width = w;
        this.computedWidth = w;
    }

    protected noteLoop(action: (note: Note) => void): void {
        for (let i: number = this.container.beat.notes.length - 1; i >= 0; i--) {
            action(this.container.beat.notes[i]);
        }
    }

    public addEffect(g: Glyph) {
        super.addGlyph(g);
        this._effectGlyphs.push(g);
    }

    public addNormal(g: Glyph) {
        super.addGlyph(g);
        this._normalGlyphs.push(g);
    }

    protected get effectElement(): BeatSubElement | undefined {
        return undefined;
    }

    public override paint(cx: number, cy: number, canvas: ICanvas): void {
        this.paintEffects(cx, cy, canvas);
        this.paintNormal(cx, cy, canvas);
    }

    private paintNormal(cx: number, cy: number, canvas: ICanvas) {
        for (const g of this._normalGlyphs) {
            g.paint(cx + this.x, cy + this.y, canvas);
        }
    }

    private paintEffects(cx: number, cy: number, canvas: ICanvas) {
        using _ = this.effectElement
            ? ElementStyleHelper.beat(canvas, this.effectElement!, this.container.beat)
            : undefined;
        for (const g of this._effectGlyphs) {
            g.paint(cx + this.x, cy + this.y, canvas);
        }
    }
}
